### ISSUE TITLE
Automatically Delete Images from Datastore

### ISSUE THREAD

**Issue by cmgorton** [forem/forem#13936]
# Automatically Delete Images from Datastore
**Note:** This was originally an RFC created by Andy Zhao. I am adding it as an issue in our public repo so that the MLH Fellows can work on it 😄

## Summary
When an article or collection (series) is deleted, their images persist in our datastore. Ideally we delete them automatically whenever the model instance is deleted.

## Motivation
By auto deleting our images, this will help with GDPR and user data ownership/deletion requests. Also, less data stored means less to pay for.

## Estimated size
S

## Design/Implementation
First, some context. Most of our images that are handled by Carrierwave, or mounted, are automatically deleted as a part of Carrierwave's implementation. The two models that do not do this are the Article and Collection (series) models, as their image columns take in any URL. Because they don't, we'll be directly sending an API call via the fog gem to delete images accordingly.

The basic steps are:

1. Add an `after_destroy_commit` callback to the models
2. Gather all the images and pass it along to a worker
3. The worker calls a service object to delete the images
4. Service object deletes the images via API call to AWS
5. 
First, the callback will be added. For example, in `app/models/article.rb`:

``` ruby 
class Article < ApplicationRecord
  IMG_MARKDOWN_REGEX = /!\[.*\]\(#{ApplicationConfig["APP_PROTOCOL"]}#{ApplicationConfig["AWS_BUCKET_NAME"]}.s3.amazonaws.com\/(uploads\/articles.*)\)/.freeze

	after_destroy_commit: :delete_images
  
  # ...
  
  private
  
  def delete_images
    image_paths = article.body_markdown.scan(IMG_MARKDOWN_REGEX).flatten
    main_image_presence = article.main_image&.scan(IMG_MARKDOWN_REGEX)
    image_paths << main_image_presence if main_image_presence # this isn't quite right but you get the idea
    DeleteImageWorker.perform_async(image_paths)
  end
end
```
We did the second step of gathering all the images in the above example.

Third, the worker will call the delete image service:

``` ruby 
# app/workers/images/delete_worker.rb

module Images
  class DeleteWorker
    include Sidekiq::Worker
    sidekiq_options queue: :low_priority, retry: 5

    def perform(image_paths)
      Images::Delete.call(image_paths)
    end
  end
end
```
Third, the service object deletes all the images:
``` ruby
# app/services/images/delete.rb

module Images
  class Delete
    attr_accessor :image_paths

    def initialize(image_paths)
      @image_paths = image_paths
    end

    def self.call(...)
      new(...).call
    end

    def call
      storage = Fog::Aws::Storage.new(
        region: ApplicationConfig["AWS_UPLOAD_REGION"],
        aws_access_key_id: ApplicationConfig["AWS_ID"],
        aws_secret_access_key: ApplicationConfig["AWS_SECRET"]
      )

      storage.delete_multiple_objects(ApplicationConfig["AWS_BUCKET_NAME"], image_paths)
      # returns an Excon::Response object, for ex:
      # <Excon::Response
      #  @data={
      #  		body: { "DeleteResult" => [
      #				{"Deleted" => {"Key" => "uploads/articles/filepath1.png"} },
      #				{"Deleted" => {"Key" => "uploads/articles/filepath2.png"} },
      #     ]},
      #			@status: 200>
    rescue => Excon::Response::BadRequest # 400 error
      # log to DataDog
    end
  end
end
```
## Definition of Done
1. An article's `main_image` and images in the body are deleted appropriately.
2. A collection's `main_image` is deleted when it's destroyed.

## Drawbacks
Like any additional processing, this will add additional code complexity to multiple delete paths.

Another drawback in performance is the fact that AWS S3 does not support getting files in a batch and allowing you to delete it in one API call. Realistically, we don't have articles that have 10+ images, but for those that do the worker may take a while to complete.

This issue does not cover deleting cached files.

## Unsolved Questions
- [SOLVED] How do we verify this implementation in a development environment? Will the developer have to get an AWS account? A: I believe you will need to have an AWS account to test this, but I have personally verified this method in production console . 
- What about cached images through Cloudinary? And Imageproxy/Openresty/Nginx?
   - This issue does not try to delete cached images. For Cloudinary, we are limited by 5000 API calls per month, and we'll need to explore a different way of handling this.
- [SOLVED] How do we test deleting cached images in a development environment? A: Testing for openresty can be done via https://github.com/forem/lfss

## Relevant Skills
- Rails backend

**Comment by squarebat** [comment]
Adding the methods given in the issue to their respected files should solve the issue. Understanding these methods also gave me a great insight into how ```fog``` gem is being used for deleting images. However, shall I just modify the files and raise a PR? Testing would be required, but I am unsure how to go about it. This [openresty testing](https://github.com/forem/lfss) link mentioned in Unsolved questions doesn't work. Not being able to verify changes in development might also be an issue.

If possible, I'd also like some insight into how Cloudinary, Imageproxy, Openresty, Nginx are being used in forem.

**Comment by maestromac** [comment]
Hey @squarebat, those tools depend on the context forem is set in. We're going to disregard Cloudinary because only DEV uses it. Forem instances rely on Imgproxy instead.

it works kind of like this 
1. User views an image on Forem.
2. The request is sent to Openresty/Nginx, where it checks if a cache of the image exists. It is served if that exists, if not...
3. (This isn't exactly correct but go along with this for now) Openresty request the original source of that image, ie AWS S3
4. Imgproxy reduces the size of the image by resizing the image.
5. The image is then cached and served by Openresty.

Regarding steps 3 & 4: it's actually a simultaneous request, in order to understand why you need to take a look at an image link on Forem.
```
https://forem.dev/images/1j5FpH4rqHiSJPxuxViX8sOfWrzruEwKJKZuoD4Af5M/s:1000:420/mb:500000/aHR0cHM6Ly9mb3Jl/bS5kZXYvcmVtb3Rl/aW1hZ2VzL3VwbG9h/ZHMvYXJ0aWNsZXMv/c2piZ3Jld2NyNmlv/azFlbzl6dDEucG5n
```
The `/images` is a proxy path where Imgeproxy resides. Imgproxy is a stand-alone tool and is its own server. its only job is to resize the image. That URL also includes the size specification (`s:1000:42`) and the original source image as the encrypted part of the URL toward the end. How this clears things up and lets me know if you need me to elaborate on anything.

-------------------------------------------------

**_but_** all those notes aside, I wonder if we can leverage Carrierwave without needing Fog. Take the following for example

```
uploader = ArticleImageUploader.new
uploader.retrieve_from_store!("a-file-name.jpeg")
uploader.remove!
```
I did that locally and was able to remove an image from the public folder. The beauty of this is, in theory, it should work with remove uploads too because our Carrierwave has all the appropriate S3 information. The potential problem here is it doesn't seem to say if it is successful or not: it returns `[:remove_versions!]` all the time even after the image is deleted.

I will need to try this on a test server first and will get back to you.

**Comment by squarebat** [comment]
@maestromac Thanks for the amazing explanation! The solution with Carrierwave seems more straightforward. However @cmgorton mentioned in the issue that Article and Collection models can't leverage it because their images take in any URL. I'll have to look into it to see why that's an issue. I still haven't to figured out how to delete cached images. Regardless, I'll try the above code and see if it works and raise a PR.

**Comment by maestromac** [comment]
@squarebat in general, there are two types of images.
1. The kind that was uploaded to the Forem platform (s3 or anywhere else the admin decides)
2. The kind that is already uploaded somewhere (ie. Imgur) and only referred to (ie. with markdown `![](some-image-link-on-imgur.com)`

It's the same thing on Github.

We're only worried about the one that was uploaded to our platform. Those won't have "any" URL, they will be in a predictable pattern. That's what `IMG_MARKDOWN_REGEX` is for; detecting those particular image patterns. That also might be the tricky part to get right because each Forem instance could be using a different source of storage but it's just another detail that can be worked out.

**Comment by squarebat** [comment]
@maestromac thanks for the clarification, that helps a lot!

**Comment by maestromac** [comment]
@squarebat yes we would definitely need to address that but you don't have to get it right immediately. Some ideas I have right now is to check Carrierwave's current configuration. If you have a look without our `carrierwave.rb` config, you''ll see that there's 3 ways it could be set. Accessing `Carrierwave.asset_host` (probably not correct) should give you proper image host location.

**Comment by squarebat** [comment]
@maestromac yes, I explored the implementation of carrierwave in the codebase to understand why and how it is being used. Based on that, I thought it would be best to use carrierwave and not fog for deleting the images, there'd be less room for error there. I did that for the articles model, the problem I'm facing is with writing tests for that. I couldn't figure out how to build an article object that would upload an image to datastore.
